src/slice.rs

```diff
@@ -629,3 +630,135 @@ macro_rules! s(
         &*&$crate::s![@parse ::std::marker::PhantomData::<$crate::Ix0>, [] $($t)*]
     };
 );
+
+/// Slicing information describing multiple mutable, disjoint slices.
+///
+/// It's unfortunate that we need `'a` and `A` to be parameters of the trait,
+/// but they're necessary until Rust supports generic associated types.
+pub trait MultiSlice<'a, A, D>
+where
+    A: 'a,
+    D: Dimension,
+{
+    /// The type of the slices created by `.multi_slice_move()`.
+    type Output;
+
+    /// Slice the raw view into multiple raw views, and dereference them.
+    ///
+    /// **Panics** if performing any individual slice panics or if the slices
+    /// are not disjoint (i.e. if they intersect).
+    fn multi_slice_move(&self, view: ArrayViewMut<'a, A, D>) -> Self::Output;
+}
+
+impl<'a, A, D> MultiSlice<'a, A, D> for ()
+where
+    A: 'a,
+    D: Dimension,
+{
+    type Output = ();
+
+    fn multi_slice_move(&self, _view: ArrayViewMut<'a, A, D>) -> Self::Output {}
+}
+
+impl<'a, A, D, Do0> MultiSlice<'a, A, D> for (SliceInfo<D::SliceArg, Do0>,)
+where
+    A: 'a,
+    D: Dimension,
+    D::SliceArg: Sized,
+    Do0: Dimension,
+{
+    type Output = (ArrayViewMut<'a, A, Do0>,);
+
+    fn multi_slice_move(&self, view: ArrayViewMut<'a, A, D>) -> Self::Output {
+        (view.slice_move(&self.0),)
+    }
+}
+
+impl<'a, A, D, Do0> MultiSlice<'a, A, D> for (&SliceInfo<D::SliceArg, Do0>,)
+where
+    A: 'a,
+    D: Dimension,
+    Do0: Dimension,
+{
+    type Output = (ArrayViewMut<'a, A, Do0>,);
+
+    fn multi_slice_move(&self, view: ArrayViewMut<'a, A, D>) -> Self::Output {
+        (view.slice_move(self.0),)
+    }
+}
+
+macro_rules! impl_multislice_tuple {
+    ($($Do:ident,)*) => {
+        impl<'a, A, D, $($Do,)*> MultiSlice<'a, A, D> for ($(SliceInfo<D::SliceArg, $Do>,)*)
+        where
+            A: 'a,
+            D: Dimension,
+            D::SliceArg: Sized,
+            $($Do: Dimension,)*
+        {
+            type Output = ($(ArrayViewMut<'a, A, $Do>,)*);
+
+            fn multi_slice_move(&self, view: ArrayViewMut<'a, A, D>) -> Self::Output {
```

Comment:
If we keep both these impls, one of them could just forward to the other, I think?

Reply:
Yeah, that's a good idea. I've made the change.